internal/manifest/ready_check.go

```diff
@@ -0,0 +1,102 @@
+package manifest
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/kyma-project/lifecycle-manager/api/v1beta2"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/cli-runtime/pkg/resource"
+	deploymentutil "k8s.io/kubectl/pkg/util/deployment"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	declarative "github.com/kyma-project/lifecycle-manager/internal/declarative/v2"
+)
+
+const customResourceStatePath = "status.state"
+
+// NewCustomResourceReadyCheck creates a readiness check that verifies that the Resource in the Manifest
+// returns the ready state, if not it returns not ready.
+func NewCustomResourceReadyCheck() *CustomResourceReadyCheck {
+	return &CustomResourceReadyCheck{}
+}
+
+type CustomResourceReadyCheck struct{}
+
+var (
+	ErrNoDeterminedState   = errors.New("could not determine state")
+	ErrCRInUnexpectedState = errors.New("module CR in unexpected state during readiness check")
+)
+
+func (c *CustomResourceReadyCheck) Run(ctx context.Context,
+	clnt declarative.Client,
+	obj declarative.Object,
+	resources []*resource.Info,
+) (declarative.StateInfo, error) {
+	if !isDeploymentReady(clnt, resources) {
+		return declarative.StateInfo{State: declarative.StateProcessing, Info: "module operator deployment is not ready"}, nil
+	}
+	manifest := obj.(*v1beta2.Manifest)
+	if manifest.Spec.Resource == nil {
+		return declarative.StateInfo{State: declarative.StateReady}, nil
+	}
+	res := manifest.Spec.Resource.DeepCopy()
+	if err := clnt.Get(ctx, client.ObjectKeyFromObject(res), res); err != nil {
+		return declarative.StateInfo{State: declarative.StateError}, err
+	}
+	stateFromCR, stateExists, err := unstructured.NestedString(res.Object,
+		strings.Split(customResourceStatePath, ".")...)
+	// CR state might not been initialized, put manifest state into processing
+	if !stateExists {
```

Comment:
is it ok to verify that before error check?

Reply:
yes, because if the CR state is not initialized, which means the `unstructured.NestedString` will return error, if we deal with error before this exist check, the manifest CR will always be in `error` first, then into `processing`